### DISCUSSION TITLE
Can HTTP/2 multi-stream speed up downloading with Hyper?

### DISCUSSION THREAD

**Discussion by OldMansKid** [hyperium/hyper#3537] [Q&A] [answered]
Hello community,
I'm implementing something like a multithreaded download manager, but using HTTP/2. From the post https://stackoverflow.com/questions/44019565, an HTTP/2 implementation needs to meet a few requirements to gain the advantage of "multi-threading" by using multi-stream.

Hyper has the initial window control function, but I wonder if Hyper interleaves frames. Or a more straightforward question, will it be faster to download large files that support range requests with multiple HTTP/2 streams?

**Comment by dswij** [comment, marked as answer]
> Or a more straightforward question, will it be faster to download large files that support range requests with multiple HTTP/2 streams?

No. And I think you have a bit of a misunderstanding on the advantage of HTTP/2. It is not `multi-threading`, but multiplexing on a single connection. Basically, it allows responses to be streamed out-of-order when making multiple requests, and this advantage can be taken of by the server (i.e. processing some requests, while also sending responses to the others).

Speed of downloading large file would mostly depend on network speed.

> but I wonder if Hyper interleaves frames

Yes, the HTTP/2 implementation of hyper sends data frames of different streams "interleaved".

**Reply by OldMansKid** [reply]
Thanks for the response. Good to know that Hyper works so well. And I do understand that HTTP/2 multiplexes frames over a single connection, but I'm not sure if opening multiple streams to download different segments of a large file would speed up, as it does in HTTP/1.1 by opening multiple TCP connections.

**Reply by dswij** [reply]
I guess opening multiple TCP connections increases the bandwidth when downloading large files, and you can achieve speed up that way.

But using HTTP/2 for this doesn't help. You still need multiple connections, not stream multiplexing provided by HTTP/2.
